File: core/num.py
from sym import *
class Num(Sym):
    tag = 'num'
    def __init__(self, V): Sym.__init__(self, V) ; self.val = float(V)
    def pfxadd(self): return self
    def pfxsub(self): return Num(-self.val)
    def add(self, o):
        if o.tag in ['int','num']: return Num(self.val + o.val)
        return Sym.add(self, o)
    def sub(self, o):
        if o.tag in ['int','num']: return Num(self.val - o.val)
        return Sym.sub(self, o)
    def mul(self, o):
        if o.tag in ['int','num']: return Num(self.val * o.val)
        return Sym.mul(self, o)
    def div(self, o):
        if o.tag in ['int','num']: return Num(self.val / o.val)
        return Sym.div(self, o)
    def pow(self, o):
        if o.tag in ['int','num']: return Num(self.val ** o.val)
        return Sym.pow(self, o)

File: core/num.cpp
#include "hpp.hpp"

Num::Num(string V):Sym("num","") { val = atof(V.c_str()); }
Num::Num(double D):Sym("num","") { val = D; }
Sym* Num::eval(Sym*E) { return this; }
string Num::head() { ostringstream os;
	os.precision(numeric_limits<double>::digits10 + 1);
	os <<"<"<< tag <<":"<< val <<"> @"<<this ; return os.str(); }
Sym* Num::pfxadd() { return this; }
Sym* Num::pfxsub() { return new Num(-val); }

Sym* Num::mul(Sym*o) {
	if (o->tag!=tag) return Sym::mul(o);
	return new Num(val * /**/ dynamic_cast<Num*>(o)->val);
}
